### PR TITLE
Issue 4531

### PR DESCRIPTION
**Pull request by RichiWIP**
### Description of the PR

Arrays are now serialized according to the different collection formats.
For collections of format "multi", all items are now appended to the parameter.
For all other formats, a string is constructed using Array.join(). The delimiter for this is read from a constant map exported by variables.ts.
All api classes are also exported in a const array to make handling of large api libraries easier.

[https://github.com/swagger-api/swagger-codegen/issues/4531](url)

### CODE REVIEW

File: samples/client/petstore/typescript-angular2/npm/api/PetApi.ts [hunk 1]
```diff
@@ -258,7 +258,7 @@ export class PetApi {
             throw new Error('Required parameter petId was null or undefined when calling deletePet.');
         }
 
-        headers.set('api_key', String(apiKey));
+            headers.set('api_key', String(apiKey));
```
Comment (wing328) [inline samples/client/petstore/typescript-angular2/npm/api/PetApi.ts]: Several additional spaces added to this line. Not sure if it's related to this PR.

File: samples/client/petstore/typescript-angular2/npm/api/PetApi.ts [hunk 2]
```diff
@@ -308,7 +308,11 @@ export class PetApi {
         let queryParameters = new URLSearchParams();
         let headers = new Headers(this.defaultHeaders.toJSON()); // https://github.com/angular/angular/issues/6845
         if (status !== undefined) {
-            queryParameters.set('status', <any>status);
+            if(status){
```
Comment (wing328) [inline samples/client/petstore/typescript-angular2/npm/api/PetApi.ts]: Can we do `if (status) {` instead so as to conform to the TS style guide: https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines#style ?

File: modules/swagger-codegen/src/main/resources/typescript-angular2/api.mustache [hunk 3]
```diff
@@ -94,12 +94,33 @@ export class {{classname}} {
 {{/allParams}}
 {{#queryParams}}
         if ({{paramName}} !== undefined) {
+            {{#isListContainer}}
+            if({{paramName}}){
+            {{#isCollectionFormatMulti}}
+                    {{paramName}}.forEach((element) => {
```
Comment (wing328) [inline modules/swagger-codegen/src/main/resources/typescript-angular2/api.mustache]: There seems to be 4 additional spaces added to the indentation.

**Comment by wing328** [comment]
@RichiWIP thanks for the PR. I've left some comments.

For `if` statement, I prefer something like the following
```
if (status) {
```
- a space before and after condition
- open curly bracket in the same line

Thanks for removing the trailing spaces in the templates.

**Comment by wing328** [comment]
@RichiWIP thanks for your PR, which looks good to me and has been merged into master.

cc @Vrolijkx

**Comment by kernwig** [comment]
This change has been lost in the v2.3.0 branch.

**Comment by wing328** [comment]
@kernwig let me cherry-pick the commits into 2.3.0 and see how it goes...

**Comment by wing328** [comment]
@kernwig I've manually reviewed and applied the fix (1 liner) via #5109. Please take a look when you've time.

**Comment by kernwig** [comment]
@wing328 I added that line and the project compiles again. I don't know see what purpose it serves though. I believe the APIS variable was originally useful in v2.2 to import and provide the services in our own modules. In v2.3, the template creates api.module.ts, which imports each service. Thus the APIS constant is now unused.

**Comment by kernwig** [comment]
Oops. No, my IDE was happy but failed at compile time. You used class name instead of file name.

Correct template line is:

`import { {{ classname }} }  from './{{ apiFilename }}';
`

**Comment by wing328** [comment]
@kernwig right, I'll file a PR to use {{apiFilename}} in both master and 2.3.0

**Comment by wing328** [comment]
@kernwig I've filed https://github.com/swagger-api/swagger-codegen/pull/5131 to fix the API file using {{classFilename}}. Please take a look to see if the change looks good to you.

I also update the line about [export](https://github.com/swagger-api/swagger-codegen/pull/5131/files#diff-976223d7122911c20d88266293d57392R4). Please review as well.

**Pull request merged**